Cleanup COLLADA support

clalancette opened the pull request:
The main goal of this series of patches is to tighten up COLLADA support, and make sure that it conforms to the specification as listed at https://www.khronos.org/files/collada_spec_1_5.pdf .

However, we also do a bunch of other fixes and improvements along the way.  This includes making sure we don't take partial updates (if the parsing fails halfway through), adding in flake8, and making joint_state_publisher always listen to the `/robot_description` topic for an updated description.

clalancette committed:
Clean up parsing to handle errors much better.

It is very possible to start parsing a URDF or
COLLADA file, and for an error to occur during that
parsing.  Previously, if this happened we would leave
the object in an undefined state with respect to the
free_joints, joint_list, and dependent_joints.

With this commit, we do all of the parsing and collecting
of data at once.  If that is completely successful, then
we update the object state to include that, so we can
no longer get partial parses.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Refactor names to be more descriptive.

The code should be a little easier to follow this way.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Additional error checking for COLLADA files.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Unindent init_collada loop.

This just makes it easier to read.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Always create the subscription to 'robot_description'.

That way it can be dynamically updated, even when the
initial configuration was given on the command-line.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

quarkytale commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -187,33 +220,39 @@ def init_urdf(self, robot):
+            else:
+                (free_joints, joint_list, dependent_joints) = self.init_urdf(xmldom)
+        except Exception as e:
+            self.get_logger().warn('Invalid robot description: %s' % (str(e)))
```
What kind of exceptions can we get here?

clalancette replied:
Good question.

It looks to me like there are a few different places that could possibly throw exceptions.  For instance, if the SDF file was missing the `model` element at line https://github.com/ros/joint_state_publisher/blob/0ea02939f79899cf1cb51edeff3d09bd62553afb/joint_state_publisher/joint_state_publisher/joint_state_publisher.py#L70 , we would get an `IndexError` exception.  Similarly, `init_collada` can throw exceptions if the COLLADA file is of the wrong version, or if there is no `kinematics_model` or `technique_common` included in the XML (https://github.com/ros/joint_state_publisher/blob/0ea02939f79899cf1cb51edeff3d09bd62553afb/joint_state_publisher/joint_state_publisher/joint_state_publisher.py#L122-L130).  Overall, anytime we have a malformed XML file (of whatever type), I want to make sure that we ignore it.

Actually, though, that brings up something I hadn't considered before.  There are actually two situations in which `configure_robot` is called.  The first situation is waiting for a valid description on the `robot_description` topic during runtime.  At that time, we definitely don't want to crash if we get an invalid XML.  The second situation where this is called is when the robot description is passed on the command-line.  If we fail parsing there, then we do actually want to quit because we can't do much meaningful, and it is better to warn the user early on.

Because of that, I think I'm going to change `configure_robot` to return a boolean, and check it only in the case it was passed on the command-line.

clalancette replied:
See 16d124666bce8dcd7991a14ef5e96218e041c574

quarkytale commented on joint_state_publisher/joint_state_publisher/joint_state_publisher.py:
```diff
@@ -99,36 +107,60 @@ def init_sdf(self, robot):
+            if joint:
+                limit = joint.getElementsByTagName('limits')[0]
+                minval = float(limit.getElementsByTagName('min')[0].childNodes[0].nodeValue)
+                maxval = float(limit.getElementsByTagName('max')[0].childNodes[0].nodeValue)
```
Should we catch `ValueError` or limits not specified perhaps?

clalancette replied:
This is another good point.

Right now, we do different things between `init_sdf`/`init_urdf` and `init_collada`.  In the former case, we actually catch the exception of a missing limit, print a warning message, skip the current joint, and continue processing the rest of the joints.  In the latter case, we currently fail the entire process, and don't update the XML at all.

I think the `init_collada` way is actually better.  That is, I think it is a lot more noticeable to users if we fail to parse the *entire* description, rather than just be missing one joint.  That latter behavior can be easy to miss and hard to diagnose.  I'm actually going to update this PR to remove the special handling of this in `init_sdf` and `init_urdf`.

clalancette replied:
See 16d124666bce8dcd7991a14ef5e96218e041c574 .

quarkytale replied:
Agreed but even if we don't want to ignore the joint and move forward, I think there could be warning messages printed in all three functions as to which joint is lacking limits so it's easier to debug and update the XML.

clalancette replied:
Yes, good point.

After looking at this for a while, I ended up getting myself quite confused.  I think the problem is that we don't have good tests which define what the various parsers do at the moment, so there is no clear way forward on where we want to go.

So I decided to take a step back and write a series of parsing tests.  These tests show what the *current* behavior of the various parsers we have are.  I'll open that as a separate PR which I think we should get in first.  Once we have that, then this PR can make modifications to both the code and the tests, and we can be a lot more confident in the direction we've taken.

clalancette replied:
And see #97 now.

clalancette committed:
Fixes from review.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette committed:
Make error messages for joint limits much more specific.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
Also see #98 , where I am adding in the linters separately.  I think we should get that one in next, and then I'll go ahead and rebase this one on top of all of that work.

clalancette committed:
Fix parsing tests with the new behavior.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
@quarkytale OK, I've now rebased this, fixed up the tests, and added all new tests.  With these changes in place, the behavior is a lot more predictable, and is very similar across URDF, SDF, and COLLADA.  Take a look when you get a chance and let me know what you think.

quarkytale reviewed (approved):
LGTM
